python_modules/dagster/dagster/components/core/defs_module.py

```diff
@@ -142,6 +143,16 @@ class DefsFolderComponentYamlSchema(Resolvable):
     asset_post_processors: Optional[Sequence[AssetPostProcessor]] = None
 
 
+@record
+class ComponentPath:
+    """Identifier for Components, targets a file and optionally an index in to that file
+    to support defining multiple Components in the same file.
+    """
+
+    file_path: Path
+    index: Optional[int] = None
```

Comment:
I think we may want `Union[str, int]` here. I'm thinking of the case where we have multiple `@component_instance` declarations in the same file and it makes sense to preserve the function name as the unique key.